library/cc/engine.cc

```diff
@@ -0,0 +1,42 @@
+#include "engine.h"
+
+#include "library/common/main_interface.h"
+#include "library/common/types/c_types.h"
+
+namespace Envoy {
+namespace Platform {
+
+Engine::Engine(envoy_engine_t engine, const std::string& configuration, LogLevel log_level,
+               std::function<void()> on_engine_running)
+    : engine_(engine), on_engine_running_(on_engine_running) {
+  envoy_engine_callbacks callbacks{
+      .on_engine_running = &Engine::c_on_engine_running,
+      .on_exit = &Engine::c_on_exit,
+      .context = this,
+  };
+
+  run_engine(this->engine_, callbacks, configuration.c_str(),
+             log_level_to_string(log_level).c_str());
+
+  this->stream_client_ = std::make_shared<StreamClient>(this->engine_);
+  this->pulse_client_ = std::make_shared<PulseClient>();
+}
+
+Engine::~Engine() { terminate_engine(this->engine_); }
+
+StreamClientSharedPtr Engine::stream_client() { return this->stream_client_; }
+PulseClientSharedPtr Engine::pulse_client() { return this->pulse_client_; }
+
+void Engine::c_on_engine_running(void* context) {
```

Comment:
I don't think it needs to be a blocker or necessarily changed here, but I might consider putting these in an anonymous namespace instead of using private static functions.

Reply:
sounds good with me. to keep this consistent it's either got to be an update across this and #1218 or a fast follow. i'd prefer the latter

Reply:
ohhh actually i remember why i did this in the first place. i wanted to scope the closures for `on_engine_running_` and `on_exit_` (if ever used) privately under `Engine`. putting the static methods under `Engine` allowed them to access those closures.

is there another pattern that would let me do that w/ an anonymous namespace?

Reply:
Anonymous namespaces are only accessible from the file in which they're declared. The header won't even need to expose that they exist. Is that what you mean?

Reply:
synced offline--going to keep as-is right now because i want to be able to access `Engine::on_engine_running_` (the `std::function`) from inside `Engine::c_on_engine_running`.

long-term vision = move the closure definition into an `EngineCallbacks`